pkg/uroot/bb.go

```diff
@@ -250,6 +254,93 @@ func (p *Package) rewriteFile(opts BuildOpts, f *ast.File) bool {
 	// Inspect the AST and change all instances of main()
 	var pos token.Position
 	hasMain := false
+
+	init := &ast.FuncDecl{
+		Name: ast.NewIdent(fmt.Sprintf("Init%d", p.initCount)),
+		Type: &ast.FuncType{
+			Params:  &ast.FieldList{},
+			Results: nil,
+		},
+		Body: &ast.BlockStmt{},
+	}
+	p.init += fmt.Sprintf("%s.Init%d()\n", p.name, p.initCount)
+	p.initCount++
+
+	// Map of fully qualified package name -> imported alias in the file.
+	imports := make(map[string]string)
+	for _, impt := range f.Imports {
+		if impt.Name != nil {
+			// TODO eval this or find a post-AST source for this info.
+			imports[impt.Path.Value[1:len(impt.Path.Value)-1]] = impt.Name.Name
+		}
+	}
+
+	// When the types.TypeString function translates package names, it uses
+	// this function to map fully qualified package paths to a local alias.
+	qualifier := func(p *types.Package) string {
+		name, ok := imports[p.Path()]
+		if ok {
+			return name
+		}
+		// TODO: is this a hack? should probably compare path against
+		// p.pkg.ImportPath?
+		if p.Name() == "main" {
+			return ""
+		}
+		return p.Name()
+	}
+
+	for _, decl := range f.Decls {
+		switch d := decl.(type) {
+		case *ast.GenDecl:
+			// We only care about vars.
+			if d.Tok != token.VAR {
+				break
+			}
+			for _, spec := range d.Specs {
+				s := spec.(*ast.ValueSpec)
+				if s.Values == nil {
+					continue
+				}
+
+				// Add an assign statement for these values to init.
+				//
+				// TODO: Double-check the order in which
```

Comment:
FYI, variables are initialized only after their dependencies: https://golang.org/ref/spec#Package_initialization

go/types.Info.InitOrder contains this ordering: https://godoc.org/go/types#Info

Reply:
To answer the TODO, init functions always come after declarations: https://github.com/golang/go/blob/master/src/cmd/compile/internal/gc/init.go#L71-L72

Reply:
:heart_eyes: thanks!